### DISCUSSION TITLE
What's the relationship between CUDA and CCCL?

### DISCUSSION THREAD

**Discussion by jiapei100** [NVIDIA/cccl#1829] [Q&A]
I got **CUDA 12.5** installed, as well as **CCCL 2.4.0** 

But, it looks there are some incompatility?

I got a couple of folders included

- `/usr/local/cuda/include`
- /usr/local/cuda/include/cuda/std/detail/libcxx/include`

But, when I tried to build **OpenCV 4.10**, I obtained the following **ERROR** messages:

```bash
/usr/local/cuda/include/cuda/std/__type_traits/enable_if.h(27): error: incomplete type "cuda::std::__4::_CCCL_VISIBILITY_DEFAULT" is not allowed
  template <bool, class _Tp = void> struct _CCCL_VISIBILITY_DEFAULT enable_if {};
                                                                    ^

/usr/local/cuda/include/cuda/std/__type_traits/enable_if.h(28): error: incomplete type "cuda::std::__4::_CCCL_VISIBILITY_DEFAULT" is not allowed
  template <class _Tp> struct _CCCL_VISIBILITY_DEFAULT enable_if<true, _Tp> {typedef _Tp type;};
                                                       ^

/usr/local/cuda/include/cuda/std/__type_traits/enable_if.h(28): error: expected an expression
  template <class _Tp> struct _CCCL_VISIBILITY_DEFAULT enable_if<true, _Tp> {typedef _Tp type;};
                                                                             ^

/usr/local/cuda/include/cuda/std/__type_traits/enable_if.h(28): error: expected a "}"
  template <class _Tp> struct _CCCL_VISIBILITY_DEFAULT enable_if<true, _Tp> {typedef _Tp type;};
                                                                                             ^

/usr/local/cuda/include/cuda/std/__type_traits/enable_if.h(30): error: name followed by "::" must be a class or namespace name
  template <bool _Bp, class _Tp = void> using __enable_if_t = typename enable_if<_Bp, _Tp>::type;
                                                                       ^

/usr/local/cuda/include/cuda/std/__type_traits/enable_if.h(33): error: name followed by "::" must be a class or namespace name
  template <bool _Bp, class _Tp = void> using enable_if_t = typename enable_if<_Bp, _Tp>::type;

```

Well, I tried to solve my compilation issue from [https://github.com/opencv/opencv_contrib/issues/3690].(https://github.com/opencv/opencv_contrib/issues/3690). However, it turns out **CCCL** is the solution .

Anyway, can anybody give me some suggestions?

Cheers
Pei

**Comment by miscco** [comment]
The error points you to an incomplete type:
> error: incomplete type "cuda::std::__4::_CCCL_VISIBILITY_DEFAULT"

However, `_CCCL_VISIBILITY_DEFAULT` is not a type but a macro that conditionally defines an attribute. The fact that the compiler thinks this is a type tells us that the macro is not defined in your build.

The only way this can happen is when you mix different versions of CCCL. Could it be that you have a separate `Thrust` dependency or pulling in CUB?

**Reply by pauleonix** [reply]
Couldn't the cause also be directly adding `/usr/local/cuda/include/cuda/std/detail/libcxx/include` to the included folders? Having `/usr/local/cuda/include` should suffice as headers in `/usr/local/cuda/include/cuda/std/detail/libcxx/include` are included from headers in `/usr/local/cuda/include/cuda/std`.

**Comment by miscco** [comment]
So `_CCCL_VISIBILITY_DEFAULT` was only introduced after 2.4 was out, so you definitely have some newer cccl repo somewhere

**Comment by jiapei100** [comment]
@miscco 
Yes, I believe I have a mixed version of CUDA or CCCL.

I have both **CUDA 12.5**  and **CCCL 2.4.0** installed under the same folder `/usr/local/cuda`.

Cheers

**Reply by jrhemstad** [reply]
@jiapei100 can you clarify if you are just simply downloading the 12.5 CUDA Toolkit and trying to use ` #include<cuda/std/tuple>` , or are you also downloading CCCL from GitHub and adding it to your include path? 

It'd be helpful if you could show your exact `nvcc` compiler invocation.

**Comment by jrhemstad** [comment]
@jiapei100 to answer the larger question of "What is the relationship between CUDA and CCCL", CCCL is included by default with the CUDA Toolkit. You can download the CUDA toolkit and include a CCCL header like `#include <cuda/std/tuple>` and it "just works".

We also support using a _newer_ version of CCCL in case you want to upgrade CCCL without upgrading the version of the CTK you're using. 

For more details, you can see https://github.com/NVIDIA/cccl?tab=readme-ov-file#platform-support
